### PR TITLE
[WIP][CELEBORN-2341] Preserve partially-committed partitions on CommitFiles timeout

### PR DESCRIPTION
**Pull request by shlomitubul**
### What changes were proposed in this pull request?

This targets `main` and supersedes #3706 (which targeted `branch-0.6`).

When `celeborn.worker.commitFiles.timeout` fires, `Controller` cancels the per-partition commit tasks (`future.cancel(true)` / `task.cancel(true)`) and runs the `handleAsync` error branch. Today that branch:

1. **never calls `context.reply()`** — the originating `CommitFiles` RPC is left unanswered. The driver only learns the outcome after its ask times out (`celeborn.client.rpc.commitFiles.askTimeout`, which falls back to `celeborn.rpc.askTimeout`, default `60s`; often raised to `240s`+), after which it re-asks with the **same commit epoch** and the worker's dedup returns the cached response;
2. **reports empty committed lists** (`COMMIT_FILE_EXCEPTION` with all requested ids failed), discarding partitions that finished closing before the timer fired;
3. **never stops the `COMMIT_FILES_TIME` timer** (latent leak);
4. **never releases the reserved slots / partition locations** at commit time (they were only reclaimed later by the shuffle-expiry cleanup in `Worker.cleanup`).

This PR makes the cancel/timeout branch:
- `context.reply(response)` immediately, so the driver doesn't wait out the ask timeout;
- `workerSource.stopTimer(COMMIT_FILES_TIME, shuffleKey)`;
- release slots and remove partition locations before replying, mirroring the normal `reply()` path (previously they were held until shuffle expiry);
- build the response from the actual commit state via a new pure helper `Controller.buildCommitFilesResponseOnCancel`, deriving the status from the failed lists materialized in the response: `SUCCESS` when nothing actually failed (every requested partition committed or closed empty before the snapshot — i.e. cancellation raced with completion), the unchanged `COMMIT_FILE_EXCEPTION` when nothing committed and nothing is empty, and `PARTIAL_SUCCESS` otherwise.

The failed lists are computed as **`requested − committed − empty`**, not just the explicitly-failed ids (see correctness note below).

### Why are the changes needed?

**Primary — eliminate the driver stall.** On every commit timeout the driver currently blocks for a full `commitFiles.askTimeout` before the same-epoch retry retrieves the worker's cached response. Replying immediately removes that wait (and the redundant retry round). The `stopTimer` call fixes a timer leak on the same path. These two are unconditional wins on every timeout.

**Secondary — make the partial-success path safe.** Tasks still queued (or interrupted before reaching a terminal state) when cancellation fires land in *none* of the worker's `committed` / `empty` / `failed` sets. If such a partition were reported as neither committed nor failed, the driver would record it nowhere; `checkDataLost` keys only off the failed sets, so it would not flag it; `collectResult` would omit it from the reducer file group; and on the read side `CelebornShuffleReader` treats a partition **absent from the file group as empty-and-valid** (it filters the requested range to `partitionGroups.containsKey(p)`), so reducers would silently produce wrong results with no `FetchFailedException`. Computing `failed = requested − committed − empty` guarantees every not-actually-committed partition is reported failed, routing it through `checkDataLost` → `SHUFFLE_DATA_LOST` → stage recompute. This is required to return `PARTIAL_SUCCESS` without regressing into silent data loss.

**Status correctness.** The driver records the worker in `commitFilesFailedWorkers` (→ `WorkerStatusTracker.excludedWorkers`) for any terminal status other than `SUCCESS`. When cancellation races with completion and the failed lists come out empty, the response is functionally identical to a success-path response, so the helper returns `SUCCESS` — otherwise the worker would be excluded from this client's slot allocations despite no actual failure. This is safe because the committed/empty sets are append-only: empty failed lists imply every requested partition had already reached a terminal good state at snapshot time.

### Scope / honest limitations

Reduce-shuffle finalization is **all-or-nothing**: `ReducePartitionCommitHandler.handleFinalCommitFiles` calls `collectResult` (which populates the reducer file groups) only when `checkDataLost` returns `false`, and `checkDataLost` flags the **whole shuffle** if any primary failed (non-replicated) or any partition failed on both replicas. So this PR does **not** let reducers fetch the committed partitions while only the lost ones recompute — when a timeout leaves any partition uncommitted in a non-replicated shuffle, the whole map stage still recomputes, exactly as before. Preserving committed partitions avoids recompute only when the timeout left nothing actually uncommitted (e.g. the timer fired as the last file closed), and otherwise just keeps the worker's report truthful. The dependable, every-time benefits are the immediate reply and the timer fix.

### Does this PR introduce any user-facing change?

No protocol or API change. `StatusCode.PARTIAL_SUCCESS` is already part of `CommitFilesResponse` and is already handled by the driver as a terminal, non-retry status (same branch as `SUCCESS` in `CommitHandler.doParallelCommitFiles`). Note that, like any non-`SUCCESS` terminal status, it records the worker in `commitFilesFailedWorkers` — which is why the helper returns `SUCCESS` when the failed lists are empty.

### How was this patch tested?

New `ControllerSuite` unit tests for `buildCommitFilesResponseOnCancel`:
- some committed, one empty, two still-queued/in-flight → `PARTIAL_SUCCESS`, committed preserved, the in-flight ids reported failed, the empty id not failed;
- nothing committed → `COMMIT_FILE_EXCEPTION` with all requested ids failed;
- nothing committed, some empty, some in-flight → `PARTIAL_SUCCESS`, empty ids not reported failed;
- all requested partitions empty → `SUCCESS` with no failures;
- every requested partition committed or empty before the snapshot (cancellation raced with completion) → `SUCCESS`, committed preserved, no failures.

Built and `spotless:check`-clean on `worker` with Java 17; all 5 suite tests pass.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by SteNicholas — COMMENTED** [review]
I traced the full data path on both the driver and reader sides — this is a solid, well-reasoned change. The two unconditional wins (immediate reply, timer-leak fix) are clearly correct, and the riskier partial-success logic is safe.

### The safety argument holds (verified)

The crucial claim is that an in-flight partition reported as *neither committed nor failed* causes silent data loss, so `failed = requested − committed − empty` is required. Every link checks out:

- `CommitHandler.checkDataLost` keys **only** off the failed sets — a partition absent from `failedPrimaryPartitionIds`/`failedReplicaPartitionIds` is never flagged.
- `collectResult` only populates the reducer file group from `committedPrimaryIds`.
- `CelebornShuffleReader` (`// filter empty partition` → `.filter(p => fileGroups.partitionGroups.containsKey(p))`) silently skips a partition absent from the file group — no `FetchFailedException`.

So an in-flight partition absent from both sets is silently dropped at read time. Computing `failed = requested − committed − empty` correctly routes it through `checkDataLost → SHUFFLE_DATA_LOST → recompute`. `PARTIAL_SUCCESS` is confirmed terminal/non-retry (same branch as `SUCCESS` in `doParallelCommitFiles`), and the "honest limitations" section matches the all-or-nothing behavior in `handleFinalCommitFiles`.

### Concurrency nuance worth a code comment

`task.cancel(true)` / `future.cancel(true)` on a `CompletableFuture` **does not interrupt** the running task — `mayInterruptIfRunning` is ignored by `CompletableFuture`. So when the timer fires, the per-partition commit tasks keep running in `commitThreadPool` while the error branch snapshots the `committed`/`empty` sets. I worked through the race and it's safe: `failedPrimaryIds` is computed from the live `committedPrimaryIds` *before* the committed-list snapshot is taken, and the committed set only grows, so a partition that commits during the window lands in **both** lists (over-reports failure → safe recompute) — never in "neither". And a partition reaches `committedIds` only after `fileWriter.close()` succeeds, so committed ⇒ durable.

This is correct but non-obvious and the ordering is load-bearing. A one-line comment noting the snapshot is best-effort (tasks may still be running since cancel doesn't interrupt) would protect against a future reorder.

### Empty-partition edge case (same as Copilot's note)

The `COMMIT_FILE_EXCEPTION` branch triggers on `committed*Ids.isEmpty`, which marks genuinely-empty partitions (tracked in `emptyFile*Ids`) as failed → unnecessary recompute in the rare "all partitions empty, none committed" case. Not a correctness bug — over-reporting failure is always safe, and it byte-for-byte preserves the prior no-commit response. If you'd rather optimize it, gate on `committed.isEmpty && empty.isEmpty`; otherwise a quick reply on the Copilot thread noting it's an intentional conservative choice would close it out.

### Minor

- Tests cover only the pure helper (the risky part — acceptable), but not the empty-only edge case above, which is exactly the Copilot scenario — worth adding.

None of this blocks merge. Nice work, and the writeup/limitations section is appreciated.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Controller.scala [hunk 1]
```diff
@@ -898,3 +905,48 @@ private[deploy] class Controller(
...
+      // is treated by the driver as empty-and-valid, which would silently drop its data.
+      val failedPrimaryIds = new jArrayList[String](primaryIds)
+      failedPrimaryIds.removeAll(committedPrimaryIds)
+      failedPrimaryIds.removeAll(emptyFilePrimaryIds)
+      val failedReplicaIds = new jArrayList[String](replicaIds)
+      failedReplicaIds.removeAll(committedReplicaIds)
+      failedReplicaIds.removeAll(emptyFileReplicaIds)
+      CommitFilesResponse(
+        StatusCode.PARTIAL_SUCCESS,
+        new jArrayList[String](committedPrimaryIds),
```
[the remark this answers is not in the document]
Comment (shlomitubul) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Controller.scala]: Good catch — fixed. The cancel branch no longer special-cases empty committed lists: `COMMIT_FILE_EXCEPTION` is now returned only when nothing committed **and** nothing is empty, and the failed list is always `requested − committed − empty`, so empty-file partitions are never reported as failed (matching the success path's treatment of empties). Added a unit test for the empty-only case.

**Comment by shlomitubul** [comment]
Thanks for the thorough trace — addressed both points in the latest push (`922eba8`):

1. **Empty-partition edge case** (also Copilot's note): the helper no longer takes a separate empty-committed branch. `failed` is always computed as `requested − committed − empty`, and `COMMIT_FILE_EXCEPTION` is returned only when nothing committed **and** nothing is empty; otherwise `PARTIAL_SUCCESS`. So genuinely-empty partitions are never reported as failed (consistent with the success path), and the "all-empty, none committed" case no longer forces a recompute. Added two unit tests: empty-files-not-failed-when-nothing-committed, and all-empty-reports-no-failures.

2. **Concurrency / cancel doesn't interrupt** — added a comment at the snapshot making the ordering explicit: `cancel(true)` doesn't interrupt a running `CompletableFuture`, the committed/empty sets are append-only, and `failed` is computed (reading `committed`) *before* the `committed` snapshot — so a partition committing mid-window lands in both lists (safe over-report), never in neither. Calling out the load-bearing ordering so a future reorder doesn't reintroduce the silent-drop.

Appreciate the review.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by SteNicholas** [comment]
Re-checked the latest two pushes.

Both points from my earlier review are addressed in `922eba8`:
1. The best-effort-cancel comment in `buildCommitFilesResponseOnCancel` documents the race precisely — `failed` is computed before the committed snapshot and the sets are append-only, so a racing commit can only land in both `failed` and `committed` (safe over-report), never in neither.
2. Empty partitions are no longer reported as failed: `COMMIT_FILE_EXCEPTION` is gated on nothing-committed-and-nothing-empty, with the empty-only unit test to match.

Also verified `59c989f`'s fix for the status/snapshot inconsistency Copilot flagged: `status` is now derived from the same committed snapshots returned in the response, so `COMMIT_FILE_EXCEPTION` with non-empty committed lists is no longer possible. The residual window (the empty sets are still read live in the status check) only degrades to a safe over-report of `failed`, consistent with the documented design.

One leftover nit: the comment "COMMIT_FILE_EXCEPTION only when nothing committed and nothing empty; empty files are a successful terminal state and must not be reported as failed." appears twice in the helper — above the committed-list snapshot and again above the status computation. Worth dropping one copy before merge.

LGTM.

**Review by SteNicholas — APPROVED** [review]

**Review by SteNicholas — CHANGES_REQUESTED** [review]
@shlomitubul, please take a look at the comments of copilot.

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Controller.scala [hunk 2]
```diff
@@ -898,3 +905,54 @@ private[deploy] class Controller(
...
+    val committedReplicaIdList = new jArrayList[String](committedReplicaIds)
+    // COMMIT_FILE_EXCEPTION only when nothing committed and nothing empty; empty files are a
+    // successful terminal state and must not be reported as failed.
+    val status =
+      if (committedPrimaryIdList.isEmpty && committedReplicaIdList.isEmpty &&
+        emptyFilePrimaryIds.isEmpty && emptyFileReplicaIds.isEmpty) {
+        StatusCode.COMMIT_FILE_EXCEPTION
+      } else {
+        StatusCode.PARTIAL_SUCCESS
+      }
```
[the remark this answers is not in the document]
Comment (shlomitubul) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Controller.scala]: Fixed in `9c3e848`. The status is now derived from the failed lists materialized in the response: `SUCCESS` when both are empty, `COMMIT_FILE_EXCEPTION` when nothing committed and nothing empty, `PARTIAL_SUCCESS` otherwise. Verified the consequence you described: `CommitHandler.doParallelCommitFiles` records the worker in `commitFilesFailedWorkers` for any terminal status other than `SUCCESS`, which flows into `WorkerStatusTracker.excludedWorkers` — so the empty-failed case did penalize the worker with no actual failure. Returning `SUCCESS` there is safe because the committed/empty sets are append-only: empty failed lists mean every requested partition had already reached a terminal good state at snapshot time, the same condition the normal `reply()` path uses. Added two unit tests (all-empty, and cancellation racing with completion).

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Controller.scala [hunk 3]
```diff
@@ -705,18 +705,25 @@ private[deploy] class Controller(
...
-                  StatusCode.COMMIT_FILE_EXCEPTION,
-                  List.empty.asJava,
-                  List.empty.asJava,
-                  primaryIds,
-                  replicaIds)
-
+                commitInfo.response = response
                 commitInfo.status = CommitInfo.COMMIT_FINISHED
               }
+              context.reply(response)
```
[the remark this answers is not in the document]
Comment (shlomitubul) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Controller.scala:739]: Done in `9c3e848` — the cancel branch now releases slots and removes partition locations before replying, mirroring `reply()`. Two corrections to the premise, for the record: (1) the cleanup was not permanently skipped before — `Worker.cleanup` releases the slots and removes the shuffle when the master reports the shuffle key expired, independent of `commitInfo.status`, so the window was "until shuffle unregistration", not forever; (2) this was pre-existing behavior on the timeout path, not introduced here — before this PR the branch also set `COMMIT_FINISHED` without cleanup, and the driver's same-epoch re-ask hit the `COMMIT_FINISHED` short-circuit which replies the cached response without releasing either. Still worth fixing for parity with `reply()`, so it's in. One caveat documented in the code: unlike `reply()`, commit tasks may still be running here (`cancel(true)` does not interrupt them); a task that fetches its location after removal logs an error and marks the partition failed, which is harmless since the response already reports every not-committed-and-not-empty partition as failed, and the expiry-time release is idempotent.

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Controller.scala [hunk 4]
```diff
@@ -705,18 +705,37 @@ private[deploy] class Controller(
                 case throwable: Throwable =>
                   logError(s"$errMsg, an unexpected exception occurred.", throwable)
               }
+              // Release slots and remove partition locations before reply, mirroring reply().
+              // Unlike reply(), commit tasks may still be running here (cancel(true) does not
+              // interrupt them): a task that has not yet fetched its location will then find it
+              // removed and mark the partition failed, which is harmless -- the response below
+              // already reports every not-committed-and-not-empty partition as failed, and the
+              // shuffle-expiry cleanup that previously reclaimed these slots is idempotent.
+              val releasePrimaryLocations =
```
[the remark this answers is not in the document]
Comment (shlomitubul) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Controller.scala:718]: Fixed in `a443eee`. Confirmed the scenario: the commit task body only `catch`es `IOException`, so a non-`IOException` thrown by a task completes the future exceptionally *without* the timer firing, and the `TimerTask` was never cancelled on the error path → it fired later with the misleading `After waiting … cancel all commit file jobs` warning (the cancels themselves were no-ops on the already-completed future/tasks, so no correctness impact). Now `timeout.cancel()` runs at the top of the error branch too, mirroring the success path.

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Controller.scala [hunk 5]
```diff
@@ -898,3 +917,59 @@ private[deploy] class Controller(
...
+      status,
+      committedPrimaryIdList,
+      committedReplicaIdList,
+      failedPrimaryIds,
+      failedReplicaIds,
+      new jHashMap[String, StorageInfo](committedPrimaryStorageInfos),
+      new jHashMap[String, StorageInfo](committedReplicaStorageInfos),
+      new jHashMap[String, RoaringBitmap](committedMapIdBitMap),
+      partitionSizeList.asScala.sum,
+      partitionSizeList.size())
```
[the remark this answers is not in the document]
Comment (shlomitubul) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Controller.scala]: Done in `a443eee` — `totalWritten`/`fileCount` are now derived from a single `partitionSizeList.asScala.toVector` snapshot. One correction for the record: `size()` on a `LinkedBlockingQueue` is O(1) (an atomic counter), so this wasn't two traversals — it was one traversal plus an O(1) read. And the practical impact was negligible: both feed approximate size-estimation metrics (`fileCount` already counts only partitions ≥ `minPartitionSizeToEstimate`), so a concurrent `add` between the two reads could at most make them momentarily off-by-one, never a correctness issue. Snapshotting once is a cheap consistency tidy-up, so it's in.

**Review by SteNicholas — COMMENTED** [review]
Code review — 3 findings (1 incomplete fix, 1 edge-case bug, 1 cleanup).

**Top issue — incomplete fix (not in diff, no inline comment):**

`checkCommitTimeout` (line 849) still builds a bare `CommitFilesResponse` with `COMMIT_FILE_EXCEPTION`, empty committed lists, and no storage infos — the same class of bug this PR fixes in the `handleAsync` exception path. When a second `CommitFiles` RPC for the same epoch waits via `shuffleCommitTime` and times out via `checkCommitTimeout`, the response discards all committed partitions and reports everything as failed with empty committed lists, causing the driver to recompute already-committed partitions. Consider reusing `buildCommitFilesResponseOnCancel` there as well (it would need access to the concurrent sets, e.g. via `CommitInfo`).

**Edge-case bug (not in diff, no inline comment):**

`InterruptedException` re-throw at line 701 exits the `match` block before the new slot-release, partition-removal, `context.reply()`, and `stopTimer` code. For every other exception type the fix works correctly, but if the `handleAsync` thread is interrupted (e.g. during worker shutdown), the caller's RPC is never replied to, slots are never released, and the timer is never stopped. Consider moving the slot-release / reply / stopTimer into a `finally` block or restructuring so `InterruptedException` also gets cleanup.

See inline comments for the third finding.

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Controller.scala [hunk 6]
```diff
@@ -898,3 +921,62 @@ private[deploy] class Controller(
     }
```
Comment (SteNicholas) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Controller.scala:921]: Cleanup (duplication): `reply()` (line 595) and this new `buildCommitFilesResponseOnCancel` method duplicate the response-assembly pattern — both snapshot the concurrent `committedPrimaryIds`/`committedReplicaIds` sets into `ArrayList`s, copy `committedPrimaryStorageInfos`/`committedReplicaStorageInfos`/`committedMapIdBitMap` maps via `new jHashMap(…)`, and derive `totalWritten`/`fileCount` from `partitionSizeList`. The only differences are (a) how failed IDs are computed (`failedIds` set vs. `primaryIds - committedIds - emptyIds`) and (b) the status-decision logic.

If a new field is added to `CommitFilesResponse`, both must be updated in lockstep — a missed update silently drops the field in one path. Consider extracting the shared snapshot-and-assemble logic into a common helper, with the differing failed-ID derivation and status logic passed in or computed by the caller.

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Controller.scala [hunk 7]
```diff
@@ -898,3 +921,62 @@ private[deploy] class Controller(
...
+    // reported as failed, so COMMIT_FILE_EXCEPTION is returned only when nothing committed
+    // and nothing is empty.
+    val status =
+      if (failedPrimaryIds.isEmpty && failedReplicaIds.isEmpty) {
+        StatusCode.SUCCESS
+      } else if (committedPrimaryIdList.isEmpty && committedReplicaIdList.isEmpty &&
+        emptyFilePrimaryIds.isEmpty && emptyFileReplicaIds.isEmpty) {
+        StatusCode.COMMIT_FILE_EXCEPTION
+      } else {
+        StatusCode.PARTIAL_SUCCESS
```
Comment (sunchao) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Controller.scala:968]: [P1] Preserve timeout exclusions for PARTIAL_SUCCESS

Returning `PARTIAL_SUCCESS` here changes the worker-exclusion lifetime. `CommitHandler` treats every non-`SUCCESS` terminal response as a failed worker, but `WorkerStatusTracker.handleHeartbeatResponse` retains `COMMIT_FILE_EXCEPTION` and `WORKER_UNRESPONSIVE` for `celeborn.client.excludedWorker.expireTimeout` while `PARTIAL_SUCCESS` falls through to removal on the next healthy application heartbeat (10s by default versus the 180s exclusion window). `ChangePartitionManager` then admits the same worker again through `workerAvailable`.

For a timeout with partial progress, this can reselect the same overloaded worker for the recomputed stage—the CELEBORN-2009 retry loop that the previous `COMMIT_FILE_EXCEPTION` status avoided. Please either retain `PARTIAL_SUCCESS` in `WorkerStatusTracker` (with a heartbeat regression test and a client-first rollout), or keep a retained failure status on this timeout path while still carrying the accurate committed/failed lists.

**Review by sunchao — COMMENTED** [review]
One additional compatibility finding from tracing the timeout response through client-side worker exclusion and reselection.
